ecosystem/typescript/sdk/src/aptos_client.ts

```diff
@@ -621,6 +621,80 @@ export class AptosClient {
     const txnHash = await this.generateSignSubmitTransaction(sender, payload, extraArgs);
     return this.waitForTransactionWithResult(txnHash, extraArgs);
   }
+
+  /**
+   * Rotate an account's auth key. After rotation, only the new private key can be used to sign txns for
+   * the account.
+   * @param forAccount Account of which the auth key will be rotated
+   * @param privateKeyBytes New private key
+   * @param extraArgs Extra args for building the transaction payload.
+   * @returns PendingTransaction
+   */
+  async rotateAuthKeyEd25519(
+    forAccount: AptosAccount,
+    privateKeyBytes: Uint8Array,
+    extraArgs?: {
+      maxGasAmount?: BCS.Uint64;
+      gasUnitPrice?: BCS.Uint64;
+      expireTimestamp?: BCS.Uint64;
+    },
+  ): Promise<Gen.PendingTransaction> {
+    const { sequence_number: sequenceNumber, authentication_key: authKey } = await this.getAccount(
+      forAccount.address(),
+    );
+
+    const helperAccount = new AptosAccount(privateKeyBytes);
+
+    const challenge = new TxnBuilderTypes.RotationProofChallenge(
+      TxnBuilderTypes.AccountAddress.CORE_CODE_ADDRESS,
+      "account",
+      "RotationProofChallenge",
+      BigInt(sequenceNumber),
+      TxnBuilderTypes.AccountAddress.fromHex(forAccount.address()),
+      new TxnBuilderTypes.AccountAddress(new HexString(authKey).toUint8Array()),
+      helperAccount.pubKey().toUint8Array(),
+    );
+
+    const challengeHex = HexString.fromUint8Array(BCS.bcsToBytes(challenge));
+
+    const proofSignedByCurrentPrivateKey = forAccount.signHexString(challengeHex);
+
+    const proofSignedByNewPrivateKey = helperAccount.signHexString(challengeHex);
+
+    const payload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
+      TxnBuilderTypes.EntryFunction.natural(
+        "0x1::account",
+        "rotate_authentication_key_ed25519",
+        [],
+        [
+          BCS.bcsSerializeBytes(proofSignedByCurrentPrivateKey.toUint8Array()),
+          BCS.bcsSerializeBytes(proofSignedByNewPrivateKey.toUint8Array()),
+          BCS.bcsSerializeBytes(forAccount.pubKey().toUint8Array()),
+          BCS.bcsSerializeBytes(helperAccount.pubKey().toUint8Array()),
+        ],
+      ),
+    );
+
+    const rawTransaction = await this.generateRawTransaction(forAccount.address(), payload, extraArgs);
+    const bcsTxn = AptosClient.generateBCSTransaction(forAccount, rawTransaction);
+    return this.submitSignedBCSTransaction(bcsTxn);
+  }
+
+  async lookupAddressByPublicKey(publicKey: MaybeHexString): Promise<HexString> {
+    const resource = await this.getAccountResource("0x1", "0x1::account::OriginatingAddress");
+
+    const {
+      address_map: { handle },
+    } = resource.data as any;
+
+    const origAddress = await this.getTableItem(handle, {
+      key_type: "address",
+      value_type: "address",
+      key: HexString.ensure(publicKey).hex(),
```

Comment:
is the public key the same as the auth key here? the key of the `OriginatingAddress` table is the auth key, which is the sha hash of the public key. but the test seems to be working so I'm guessing public_key == auth_key here? or am I missing something

Reply:
Good catch, I should not call it publicKey. It is indeed `authKey`. The test actually passed in address.